### ISSUE TITLE
Merge train

### ISSUE THREAD

**Issue by ReinierMaas** [channable/hoff#77]
### Before (without merge trains), assuming build "passes"

![no-merge-trains](https://user-images.githubusercontent.com/3999598/185108707-ee495355-bc52-4b76-ad1c-5b213c7a0f07.svg)

### Objective (with merge trains), assuming build "passes"

![merge-trains](https://user-images.githubusercontent.com/3999598/185108703-74d92b05-5333-49a6-9dc4-39815d720b0c.svg)

------

Our current merge strategy exists of a single pull request being rebased on master and pushed to the testing branch. With the current setup, we are gated on the runtime of our CI pipeline. With a CI pipeline of 10 minutes at most 6 PRs can be merged in an hour.

There exists multiple ways to merge PRs more quickly with a single testing branch.

---

We can create a rebase train, the example has a length of 3 train cars:

```
testing: master<-(testing-pr1)<-(testing-pr2)<-(testing-pr3)
approved: waiting-pr4, waiting-pr5, waiting-pr6, waiting-pr7
```

Assuming that PRs that are approved are expected to merge this allows the length of the train to be merged in quick succession. When a PR fails in the middle everything in front of the train can continue to be merged into master, everything after that PR is rebased upon the latest non-failed PR. When `testing-pr2` fails the state will be updated to:

```
testing: master<-(testing-pr1)<-(testing-pr3)<-(testing-pr4)
approved: waiting-pr5, waiting-pr6, waiting-pr7
failed: failed-pr2, sets notification on the pull request
```

When `testing-pr3` successfully completes its CI run its internal state is updated:

```
testing: master<-(testing-pr1)<-success-pr3<-(testing-pr4)
approved: waiting-pr5, waiting-pr6, waiting-pr7
```

When `testing-pr1` successfully complete their CI runs its internal state is updated and the master branch is advanced to the first PR with the testing state, which would be `testing-pr4`. The state is updated to:

```
testing: master<-(testing-pr4)<-(testing-pr5)<-(testing-pr6)
approved: waiting-pr7
```

This ensures that every pr is tested in isolation before being merged into master. We would have the same confidence as we have now.

---

Another solution we can pursue is working with rollups, the example has rollup batches of 3 pull requests. Pull requests are added to the next rollup when they are approved. If just a single PR is approved that single pr will be tested in isolation, when there are multiple pull requests waiting to be merged into master then at most batch size PRs will be rebased on master.

```
testing: master<-(testing-pr1<-testing-pr2<-testing-pr3)
next batch: waiting-pr4, waiting-pr5, waiting-pr6
approved: waiting-pr7
```

When the batch completed successfully master is advanced to the latest of the testing prs commit and the next batch is scheduled:

```
testing: master<-(testing-pr4<-testing-pr5<-testing-pr6)
next batch: waiting-pr7
```

When the batch failed each pull requests is tested in isolation:

```
testing: master<-(testing-pr4)
possible failures: maybe-failed-pr5, maybe-failed-pr6
next batch: waiting-pr7
```

If the pr-4 wasn't the failure we can forward master to pr4 without a problem:

```
testing: master<-(testing-pr5)
possible failures: maybe-failed-pr6
next batch: waiting-pr7
```

Pull requests 5 is tested and results in a failure:

```
testing: master<-(testing-pr6)
next batch: waiting-pr7
failed: failed-pr5, sets notification on the pull request
```

All remaining PRs of the failed batch get merged and the next batch is started, this contains solely pr-7:

```
testing: master<-(testing-pr7)
```

---

Future speed improvements can combine the two techniques, i.e. using the train car technique but instead of having a single pr per train car, we can have batches per train car. Connecting the next batch when a batch is full, i.e. there are three pending pull requests, or when the train is empty, the train has zero cars. This would still have a limit on the maximum length of the train:

```
testing: master<-(testing-pr1<-testing-pr2<-testing-pr3)<-(testing-pr4<-testing-pr5<-testing-pr6)
next_batch: waiting-pr7
```

More incoming pull requests 8-13, we attach the third batch to the train:

```
testing: master<-(testing-pr1<-testing-pr2<-testing-pr3)<-(testing-pr4<-testing-pr5<-testing-pr6)<-(testing-pr7<-testing-pr8<-testing-pr9)
next_batch: waiting-pr10, waiting-pr11, waiting-pr12
approved: waiting-pr13
```

If the second batch of pull requests failed, we add the separate pull requests of that batch to the train and pause everything that can't currently be connected, the first batch continues with the CI run it already had:

```
testing: master<-(testing-pr1<-testing-pr2<-testing-pr3)<-(testing-pr4)<-(testing-pr5)
possible failure: maybe-failed-pr6
paused batches: (testing-pr7<-testing-pr8<-testing-pr9)
next_batch: waiting-pr10, waiting-pr11, waiting-pr12
approved: waiting-pr13
```

## TODO list

- [x] ~record parents of integrations #130~
- [x] ~replace global candidate by global list of candidates #130~
- [x] #131
    - [x] remove global candidate
    - [x] https://github.com/channable/devops/pull/5701
    - [x] #141
    - [x] https://github.com/channable/devops/pull/5738
    - [x] #144
    - [x] https://github.com/channable/devops/pull/5740
- [x] #134
    - [x] use `testing/<id>` as the testing branch
    - [x] cleanup unused testing branches
    - [x] fix issue of pushing `testing/<id>` in the presence of `testing`
    - [x] one last full test just to be safe (test upgrade)
    - [x] create message with [manual fix instructions](https://github.com/channable/hoff/pull/134#issuecomment-1190699109)
    - [x] https://github.com/channable/devops/pull/5752
- [x] #146 
    - [x] split `BuildPending` into `BuildStarted`
    - [x] slightly change automated comment wording
    - [x] include link to commit in web interface
    - [x] include screenshots in the PR
    - [x] https://github.com/channable/devops/pull/5768
- [x] #150
    - [x] #154 
    - [x] https://github.com/channable/devops/pull/5791
- [x] #155
- [ ] #137
- [ ] full test using local instance and a git-sandbox repo
    - [ ] prepare a git-sandbox repo with 2-3 minutes of artificial CI delay
    - [ ] test merging a single PR
    - [ ] test merging 2 passing PRs
    - [ ] test other pass/fail workflows
    - [ ] test [rebase failures](https://github.com/channable/hoff/issues/77#issuecomment-1178980677)
- [ ] deploy
    - [ ] repeat tests on a git-sandbox repo throughly
   - [ ] remove 2-3 minutes artificial delay if needed after
- [ ] Test using PRs in the Hoff repository itself.  Ask Hoff to merge the following PRs in the following order a few seconds apart:
   - [ ] #157 (should build and merge automatically)
   - [ ] #156 (potential conflict)
   - [ ] #158 (should build and merge automatically)
- [ ] add a configurable limit to the size of the train

**Comment by wesleybowman** [comment]
Situation for RM right now:

![Jun24::110149](https://user-images.githubusercontent.com/3688009/123234888-a5d41c80-d4db-11eb-9769-e1eefd673b68.png)



Each build takes around 20 min or longer, so....

**Comment by ruuda** [comment]
> When the batch failed each pull requests is tested in isolation:

If the batch is large, you can do even better with a binary search: try merging the full batch, if it fails, try half of the batch. If that succeeded the other half must contain a bad PR, if it failed, the first half must contain a bad PR, and then you can recurse. If CI resources are not a constraint, you can also test all options in parallel.

**Comment by ReinierMaas** [comment]
After some discussion, the decision is made to go for the rebase train idea which speculatively executes multiple pull requests. In order to limit CI runtime, we will make the setup accept a list of branch names which it will use as a ring buffer, adding more branch names to the list allows Hoff to speculatively execute more PRs concurrently.

We chose this option because it allows us to run the integration test for each pull request that is merged. This gives us the same security as merging all PRs one at a time but allows the CI execution times to overlap, this speeds up merges when the PRs CI runs are successful. It will rebase all PRs that are executing speculatively after any failed PR. This option uses more CI time in favour of quicker merges to `master`.

We expect that creating a rebase train allows quicker merges to `master`. If actions you take upon pushes to `master` cannot be executed in quick succession you should ensure external rate limits or queues for those jobs. For example, automated rollouts of the latest `master` might not be finished before the `master` branch is updated again.

We also want to thank Ruud for voicing his ideas.

**Comment by rudymatela** [comment]
Now, consider the same situation, but now there is PR `#49` which received a "merge" command after `#48`.  The idea, is to attach it to the merge train anyway:

![train-rebase-failure-2](https://user-images.githubusercontent.com/3999598/178003610-9a919e34-fe54-41a1-850f-1e05fd88dbdc.svg)

Ultimately, for PR `#48`, we only depend on the result of `testing1` and `testing2`, and not on `testing3`.  So we have four cases again:

* `testing1`:x:; `testing2`:x:: PR `#48` is rebased on top of master, if rebase suceeds, it becomes the first wagon of a new merge train which may contain PR `#49` right after it;
* `testing1`:heavy_check_mark:; `testing2`:heavy_check_mark:: `master` is eventually updated to point to `testing2`, PR `#48` receives a comment by Hoff, in the lines of: "rebase on top of master failed, please rebase manually";  The build of `#49` carries on and:
    - `testing3`:x:: it is attached at the end of the train after a fix and a new "merge" command;
    - `testing3`:heavy_check_mark:: it becomes the new master overtaking PR `#48`;
* `testing1`:heavy_check_mark:; `testing2`:x:: `master` is updated to point to `testing1`, PR `#46` receives a comment that the build failed and PR `#48` is rebased on top of the new `master` to be the new `testing1` branch; the build of `testing3` is now irrelevant, as it is built on top of something that fails;
* `testing1`:x:; `testing2` :heavy_check_mark:: we treat as if `testing1`:x:; `testing2`:x::, if a wagon fails, all wagons after it fail.  See above (and the PR description by @ReinierMaas).

In summary, when rebase fails for a given PR, we wait until the results of the train at-the-point-on-which-appending-was-tried to decide what do to with it: try to automatically rebase-then-merge again in case of build failure, ask the user to rebase manually in case of build success.

This comment is written as a sort of "non-conflicting refinement" over the above.

Thanks again for @ReinierMaas for the offline discussion.  :-)

**Comment by ruuda** [comment]
If I understand correctly, the problem is that when a rebase fails, we don’t want to tell the user (yet) to rebase manually, because the thing we tried to rebase it on top of might end up not being merged, and maybe rebasing succeeds at a later point. Or in other words, we should not cause side effects during speculative execution.\*

So here’s an idea: if the PR that fails to rebase is at the front of the train (i.e. we try to rebase it directly on top of master and it fails), notify the user. If it is not at the front of the train, that means it is speculative. If it fails to rebase we just discard it for now and continue trying to build the train with other pending PRs. CI should eventually report a build status for that train, so we should make progress. Eventually PRs will either be merged or marked as failing. So if the train is ordered by lowest PR id first, then eventually this PR that fails to rebase will be at the front of the train, and at that point we can notify the user.

I think this is a bit simpler than trying to keep track of what train we tried to rebase the PR onto. We can express the function that builds the train as a pure function `current tip -> list of pending PRs -> (new tip, list of included PRs)`. (There still needs to be _some_ new state though, that tracks once a build fails that one of the PRs part of the train must be bad, so the next iteration doesn’t build exactly the same train.)

\* <sub>Something Intel learned the hard way</sub>

**Comment by ruuda** [comment]
## Train selection idea

Okay now I nerd-sniped myself on this, so some thoughts on how to build the train. Some assumptions that are false in practice but help to simplify the problem:

* We can only test one train at a time.\* Or in other words, this is a game that proceeds in rounds. We get to pick a particular train of PRs, and either the entire train gets merged, or CI says “no”. <del>Our goal is to merge all PRs in the minimal number of rounds.</del><ins>No, our goal is to minimize the total time that PRs spend waiting.</ins>
* Failure is a property of a PR, there are no interactions between PRs. (I.e. no A+B together fail but either one on their own would succeed, and no A+B succeed but either one on their own would fail.)

\* <sub>This is not what [this comment](https://github.com/channable/hoff/issues/77#issuecomment-1067013218) says. But I think parallelism can be bolted on later; either way you need to select how you want to build the train.</sub>

<details>
<summary>Some assumptions that at first I wanted to make, but after I wrote down the idea below, these assumptions are not needed</summary>

* The set of PRs is static, there are no incoming PRs. — No actually, with a “is-good” probability per PR, this should be fine. New PRs come in with a high is-good probability and they just get included in new trains.
* In a failing train, there is at most one bad PR. — No actually, I am just working with an “is-good” probability per PR now, and it can become high for more than one PR that was part of a train.

</details>

This situation reminds me a bit of [3Blue1Brown’s Solving Wordle using Information Theory](https://www.youtube.com/watch?v=v68zYyaEmEA). We are playing this game that proceeds in rounds, we can play a subset of the set of approved PRs, and after the turn we learn something about those PRs that we included. Only this time we are not _just_ trying to maximize what we learn, because learning that a train is good is better than learning that a train contains a bad PR.

Here’s one idea. Per PR, we track the probability that it is good. Let’s say that we set it to 0.9 to start with, and it resets to that if the tip of its branch changes. When we build a train of PRs that contains PR X, and that train fails, we can do a Bayesian update of X's is-good probability.

$$
\mathbb{P}(X \textup{ is good }|\textup{ train failed}) = \frac{
  \mathbb{P}(\textup{train fails } | X \textup{ is good}) \mathbb{P}(X \textup{ is good })
}{\mathbb{P}(\textup{train fails})}
$$

This right-hand side we can compute: $\mathbb{P}(X \textup{ is good })$ we track;

$$
\mathbb{P}(\textup{train fails}) = 1 - \mathbb{P}(\textup{entire train is good}) = 1 - \prod_{Y \in \textup{train}} \mathbb{P}(Y \textup{ is good })
$$

and

$$
\mathbb{P}(\textup{train fails } | X \textup{ is good}) = 1 - \prod_{Y \neq X} \mathbb{P}(Y \textup{ is good })
$$

(almost the same, but we set the probability that X is good to 1 in this product).

Armed with this information, for any train we can estimate the probability that it succeeds. Now we have two opposing forces: if we include more PRs, the probability that the train succeeds goes down, but if it does succeed, we merge more PRs in one go. I think what we want to do here is maximize the expected number of PRs that get merged. That number is also easy to compute, it is

$$
\mathbb{E}(\textup{number of PRs merged}) = \|\textup{train}\| \prod_{Y \in \textup{train}} \mathbb{P}(Y \textup{ is good })
$$

where $\|\textup{train}\|$ is the size of the train. This is easy to maximize with a greedy algorithm, because every PR contributes equally to increasing the size of the train, so we just sort the PRs by is-good probability, and then try larger and larger trains until the expected number of PRs merged no longer increases. (I would still merge them in PR id order, but for deciding which PRs to include, we can order them by is-good probability.) Now we are done!

There might be some properties that we want a train selection algorithm to have. For example, we want it to make progress, we don’t want it to select the same train twice in a row. But when I thought more deeply about it ... when we are working with probabilities and treat build results as Bayesian evidence, I think we _do_ want to try the same train multiple times in some cases. The build could have failed due to a flake, it might pass a second time. And while it would build the same train, the is-good probabilities would go down every time, until they get so low that the full train no longer maximizes the expected number of PRs merged.

## Example

Would it work in practice? Let’s run through an example. Let’s say we have PRs A, B, C, D, E, and C is bad. After three tries, PRs F and G come in. Without speculation, this would take 7 steps to merge. How long does it take the above algorithm, with an initial is-good probability of 0.9?

**Step 1**
Probabilities: `A: 0.9, B: 0.9, C: 0.9, D: 0.9, E: 0.9`.
Trains: `A: 0.9, AB: 1.62, ABC: 2.19, ABCD: 2.6244, ABCDE: 2.95`. So we include all PRs. It fails.
The update for all PRs is the same: `P(is_good) = (1 - 0.9^4)*0.9 / (1 - 0.9^5) = 0.76`.

**Step 2**
Probabilities: `A: 0.76, B: 0.76, C: 0.76, D: 0.76, E: 0.76`.
Trains: `A: 0.76, AB: 1.16, ABC: 1.32, ABCD: 1.34, ABCDE: 1.27`. So we try ABCD. It fails.
The update for PRs A–D is the same: `P(is_good) = (1 - 0.76^3) * 0.76 / (1 - 0.76^4) = 0.64`.

**Step 3**
Probabilities: `E: 0.76, A: 0.64, B: 0.64, C: 0.64, D: 0.64`.
Trains: `E: 0.76, EA: 0.97, EAB: 0.93, EABC: 0.80, EABCD: 0.64`. So we try EA. It succeeds.
There is no Bayesian update, because all PRs that we learned something about are now out of the way. PRs F and G come in.

**Step 4**
Probabilities: `F: 0.9, G: 0.9, B: 0.64, C: 0.64, D: 0.64`.
Trains: `F: 0.9, FG: 1.62, FGB: 1.56, FGBC: 1.33, FGBCD: 1.06`. So we try FG. It succeeds.

**Step 5**
Probabilities: `B: 0.64, C: 0.64, D: 0.64`.
Trains: `B: 0.64, BC: 0.82, BCD: 0.79`. So we try BC. It fails.
The update for B and C is the same: `P(is_good) = (1 - 0.64) * 0.64 / (1 - 0.64^2) = 0.39`.

**Step 6**
Probabilities: `D: 0.64, B: 0.39, C: 0.39`.
Trains: `D: 0.64, DB: 0.50, DBC: 0.29`. So we try D. It succeeds.

**Step 7**
Probabilities: `B: 0.39, C: 0.39`.
Trains: `B: 0.39, BC: 0.30`. So we try B. It succeeds.

**Step 8**
Probabilities: `C: 0.39`.
Trains: `C: 0.39`. So we try C. It fails.

### Analysis

So it does take 7 steps as well to get everything merged in this case, but it does depend on what initial probability we choose. We can make it more conservative or more aggressive when trying to merge PRs by tweaking that initial probability.

Also, in a dynamic system, the number of steps is not what we try to minimize. What we should minimize is wait-steps.\* When we count that, in the linear example we have 5 (A–E) + 4 (B–E) + 3 (C–E) + 4 (D–G) + 3 (E–G) + 2 (F–G) + 1 (G) = 22 wait-steps.
With the above algorithm, we have 5 (A–E) + 5 (A–E) + 5 (FGBCD) + 3 (BCD) + 3 (DBC) + 2 (BC) + 1 (C) = 24 wait-steps. Hmm, my algorithm does worse on that aspect too!

I have a feeling that the algorithm is on the right track though, I only worked through one example, maybe it was unlucky. Also, the P=0.9 of a PR being good was unrealistic, as only 6/7 = 0.86 of the PRs was good. What happens when we put that in as initial probability? It would be nice to make a simulation.

\* I didn’t realize it at first, but this is what the algorithm minimizes! It maximizes the expected number of PRs it will merge, so it minimizes the wait-steps incurred by those left behind.

## Parallelism

I think the above framing is also a good one to think about parallelism. It’s a bit easier to reason about when we keep the game in discrete rounds, but now per round you can propose _n_ trains instead of a single train, and you learn the _n_ outcomes at the end of the round. This is not how builds work in practice, but maybe when we figure out how to do this, we can extend it to something more like when you have to play a train every round, but you only learn the outcome _n_ rounds later. Or maybe we should actually look at that one :thinking: Maybe in another comment.

**Comment by ruuda** [comment]
## Parallel extension of the above algorithm

After thinking about it some more, there is also a natural solution to the parallelism problem.

Let’s say train _n_ is currently building, and some build capacity becomes available. Maybe new PRs have come in in the meantime, maybe not. What train should we build?

There are two possible outcomes of build _n_:

1. It succeeds. Then it gets merged, and the next thing has to be rebased on top of this. We already have an estimate for the probability that it succeeds: the product of the is-good probabilities of the PRs in the train. Let’s call it `P(success(n))`.
2. It fails. In that case we update the is-good probabilities of the PRs in the train, and start over on top of master.

The insight is: we can build a train that maximizes the expected number of PRs merged for both situations using the earlier algorithm. And then to pick between the two, we use the one that maximizes the total expected number of PRs merged:

1. On success of train n, we expect to merge `len(train(n)) + expected_len(train(n+1))` PRs. This happens with probability `P(success(n))`. When train n fails, our work on top of does not contribute to any PRs getting merged. So the total expected number of PRs merged is `P(success(n)) * (len(train(n)) + expected_len(train(n+1)))`.
2. On success of train n, our work does not contribute to any PRs getting merged. On failure of train n, we expect to merge `expected_len(n+1))` PRs. So the total expected number of PRs merged is `(1 - P(success(n))) * expected_len(train(n+1))`.

And from these two, we select the one that produces the highest expected number of PRs merged.

An interesting consequence is that the algorithm can explore in two directions, “optimistically” and “pessimistically”. If it is pretty confident that train n will get merged, it goes the “optimistic concurrency” route and builds on top of it assuming it will pass. But when it estimates that train n has only a low probability of getting merged (for example, because it is a single PR that has been in a number of failing trains, and is almost certainly the cause of them failing, while new fresh PRs have come in that will likely pass), it will “hedge its bets” and try an alternative.

So this settles the case for when one build is running, and we want to build a train for the next build. What if two builds are running already? Or _m_ builds? We can take the same approach, but now we have 2<sup>m</sup> starting scenarios to consider. (Every existing build can either succeed or fail.) Some of these scenarios are equivalent for building on top of. For example, if train _n_ + 1 competes with (doesn’t build upon) train _n_, then when train _n_ succeeds, the outcome of train _n_ + 1 is irrelevant because we should build upon train _n_ anyway. Some of these scenarios would produce trains that are already being built. For example, if train _n_ + 1 is competing with train _n_, we still have the option to build upon train _n_ this time, and if no new PRs or build results came in, we would select the same train as _n_ + 1. In that case we should just discard it and look at the next best option.

**Comment by rudymatela** [comment]
My initial code changes were:

* the "candidate" maybe is now a "candidates" list
* the integration status includes the parent PR

Quite a lot of code was affected, ~150 lines!

I am not sure whether these are needed, but I had to start with something at least to learn what is affected.  This exploration is stored in PR #130.

I am now thinking of starting over by dropping the candidate list altogether and simply infer this information from the collection of PRs.  We'll see how that goes...

**Comment by rudymatela** [comment]
Note to self: in the code "candidate" can mean two different (but related things):

* a candidate for integration, as returned by the function `candidatePullRequests`;
* something that has been integrated and is waiting for build results to become master `integrationCandidate` field of `ProjectState`;

If I understand correctly, the latter is perhaps better described as a PR that is "integrated-but-waiting-for-build-results" or an `integratedCandidate`.

(This is relevant to the changes in #131.)

**Comment by rudymatela** [comment]
Here is an example of the current implementation (#137) running locally:

## Test Case 1, success, success, success

### Web before approval

![0-web](https://user-images.githubusercontent.com/3999598/184864037-e7a4e301-030d-4eda-b624-5b060b68cb9b.png)

### Web right after approval

![1-web](https://user-images.githubusercontent.com/3999598/184864931-130da2e2-d9e5-44c9-b7dd-e41a25e976d4.png)

### Web final

![z-web](https://user-images.githubusercontent.com/3999598/184866572-fee686a0-1338-41f8-80dc-3e3c7105dc4f.png)


### PR 120
![120](https://user-images.githubusercontent.com/3999598/184866145-db22b6d2-ecbd-4783-a41f-e41b2f2d837d.png)

### PR 121
![121](https://user-images.githubusercontent.com/3999598/184866194-23f29183-367b-4bc6-87d1-ce48c8de9a0d.png)

### PR 122
![122](https://user-images.githubusercontent.com/3999598/184866248-2469e5ac-dfa7-4d92-989c-a7838eb4c28f.png)

**Comment by rudymatela** [comment]
Here is the draft of the version announcement.

# OpsBotPrime (Hoff) v0.27.0, merge trains / speculative rebases beta

*TL;DR: OpsBotPrime now does **concurrent builds**; it **may be unstable** for the next few days, keep an eye for **anything unusual** and **report**.*

Starting with v0.27.0 deployed this morning, Hoff will perform speculative rebases to form merge trains.  If two `@OpsBotPrime merge` commands are issued in quick succession on two different PRs, the second PR will be *tentatively* rebased and built on top of the first.  You will see something like the following message:

> Speculatively rebased as [`735ab12`](https://github.com/channable/git-sandbox/commit/935ab1217fca7f74b02e7c8cc0d69540ac3947d6) behind https://github.com/channable/git-sandbox/pull/124, waiting for CI …

If the build passes on the first PR, it is merged as usual.  Since the build is already running on the second PR, it can be merged as soon as it passes as well.  If the build fails at some point in the chain the train is restarted with the next approved PR.

You can find more details in [hoff#77](https://github.com/channable/hoff/issues/77).

This was quite an intricate change in terms of the internal logic of Hoff.  **So we ask everyone to keep an eye for anything unusual being done by Hoff** and report if something appears.  This is a beta version.

The goal here is to make merges happen faster.

### Before

![before](https://user-images.githubusercontent.com/3999598/185108707-ee495355-bc52-4b76-ad1c-5b213c7a0f07.svg)

### After

![after](https://user-images.githubusercontent.com/3999598/185108703-74d92b05-5333-49a6-9dc4-39815d720b0c.svg)
